index.js

```diff
@@ -502,6 +506,455 @@
+  S.equals = def('equals', {}, [a, b, $.Boolean], Z.equals);
+
+  //# concat :: Semigroup a => a -> a -> a
+  //.
+  //. Curried version of [`Z.concat`][].
+  //.
+  //. ```javascript
+  //. > S.concat('abc', 'def')
+  //. 'abcdef'
+  //.
+  //. > S.concat([1, 2, 3], [4, 5, 6])
+  //. [1, 2, 3, 4, 5, 6]
+  //.
+  //. > S.concat({x: 1, y: 2}, {y: 3, z: 4})
+  //. {x: 1, y: 3, z: 4}
+  //.
+  //. > S.concat(S.Just([1, 2, 3]), S.Just([4, 5, 6]))
+  //. Just([1, 2, 3, 4, 5, 6])
+  //. ```
+  S.concat = def('concat', {a: [Z.Semigroup]}, [a, a, a], Z.concat);
+
+  //# empty :: Monoid a => TypeRep a -> a
+  //.
+  //. [Type-safe][sanctuary-def] version of [`Z.empty`][].
+  //.
+  //. ```javascript
+  //. > S.empty(String)
+  //. ''
+  //.
+  //. > S.empty(Array)
+  //. []
+  //.
+  //. > S.empty(Object)
+  //. {}
+  //. ```
+  S.empty = def('empty', {a: [Z.Monoid]}, [TypeRep(a), a], Z.empty);
+
+  //# map :: Functor f => (a -> b) -> f a -> f b
+  //.
+  //. Curried version of [`Z.map`][].
+  //.
+  //. ```javascript
+  //. > S.map(Math.sqrt, [1, 4, 9])
+  //. [1, 2, 3]
+  //.
+  //. > S.map(Math.sqrt, {x: 1, y: 4, z: 9})
+  //. {x: 1, y: 2, z: 3}
+  //.
+  //. > S.map(Math.sqrt, s => s.length)('Sanctuary')
+  //. 3
+  //.
+  //. > S.map(Math.sqrt, S.Just(9))
+  //. Just(3)
+  //.
+  //. > S.map(Math.sqrt, S.Right(9))
+  //. Right(3)
+  //. ```
+  S.map = def('map', {f: [Z.Functor]}, [Fn(a, b), f(a), f(b)], Z.map);
+
+  //# bimap :: Bifunctor f => (a -> b) -> (c -> d) -> f a c -> f b d
+  //.
+  //. Curried version of [`Z.bimap`][].
+  //.
+  //. ```javascript
+  //. > S.bimap(S.toUpper, Math.sqrt, S.Left('foo'))
+  //. Left('FOO')
+  //.
+  //. > S.bimap(S.toUpper, Math.sqrt, S.Right(64))
+  //. Right(8)
+  //. ```
+  S.bimap =
+  def('bimap',
+      {p: [Z.Bifunctor]},
+      [Fn(a, b), Fn(c, d), p(a, c), p(b, d)],
+      Z.bimap);
+
+  //# promap :: Profunctor p => (a -> b) -> (c -> d) -> p b c -> p a d
+  //.
+  //. Curried version of [`Z.promap`][].
+  //.
+  //. ```javascript
+  //. > S.promap(Math.abs, S.inc, Math.sqrt)(-100)
+  //. 11
+  //. ```
+  S.promap =
+  def('promap',
+      {p: [Z.Profunctor]},
+      [Fn(a, b), Fn(c, d), p(b, c), p(a, d)],
+      Z.promap);
+
+  //# alt :: Alt f => f a -> f a -> f a
+  //.
+  //. Curried version of [`Z.alt`][].
+  //.
+  //. ```javascript
+  //. > S.alt(S.Nothing, S.Just(1))
+  //. Just(1)
+  //.
+  //. > S.alt(S.Just(2), S.Just(3))
+  //. Just(2)
+  //.
+  //. > S.alt(S.Left('X'), S.Right(1))
+  //. Right(1)
+  //.
+  //. > S.alt(S.Right(2), S.Right(3))
+  //. Right(2)
+  //. ```
+  S.alt = def('alt', {f: [Z.Alt]}, [f(a), f(a), f(a)], Z.alt);
+
+  //# zero :: Plus f => TypeRep f -> f a
+  //.
+  //. [Type-safe][sanctuary-def] version of [`Z.zero`][].
+  //.
+  //. ```javascript
+  //. > S.zero(Array)
+  //. []
+  //.
+  //. > S.zero(Object)
+  //. {}
+  //.
+  //. > S.zero(S.Maybe)
+  //. Nothing
+  //. ```
+  S.zero =
+  def('zero', {f: [Z.Plus]}, [TypeRep($.TypeVariable('f')), f(a)], Z.zero);
```

Comment:
`f :: Type -> Type` is not a member of sanctuary-def's internal Type type. Should it be? I don't know. For now, at least, we need to provide a value of type `Type`.